### PR TITLE
[+] Search - Add the search on ObjectID and Number fields (exact match only)

### PR DESCRIPTION
**Pull request by adriguy**

### CODE REVIEW

File: src/services/search-builder.js [hunk 1]
```diff
@@ -19,9 +19,6 @@ function SearchBuilder(model, opts, params, searchFields) {
       fieldsSearched.push(fieldName);
     }
 
-    if (new RegExp('^[0-9a-fA-F]{24}$').test(params.search)) {
-      return { _id: params.search };
-    }
```
Comment (adriguy) [inline src/services/search-builder.js]: The search by ObjectID was broken because of this code.

File: src/services/search-builder.js [hunk 2]
```diff
@@ -30,13 +27,17 @@ function SearchBuilder(model, opts, params, searchFields) {
...
+        if (new RegExp('^[0-9a-fA-F]{24}$').test(params.search)) {
           condition[key] = opts.mongoose.Types.ObjectId(params.search);
           pushCondition(condition, key);
-        } catch (error) {
-          console.error(error);
+        }
+      } else if (value.instance === 'Number') {
+        if (!Number.isNaN(searchNumber)) {
+          condition[key] = Number(params.search);
+          pushCondition(condition, key);
```
Comment (adriguy) [inline src/services/search-builder.js]: With this additional code, we are able to search on a Number field (it has to match exactly the number - same behaviour than the one for the ObjectID).

**Comment by adriguy** [comment]
[Search - search on id field using express-mongoose returns no result](https://trello.com/c/9lEhdhdF/3779-search-search-on-id-field-using-express-mongoose-returns-no-result)

File: src/services/search-builder.js [hunk 3]
```diff
@@ -30,13 +27,17 @@ function SearchBuilder(model, opts, params, searchFields) {
...
       if (value.instance === 'ObjectID') {
-        try {
+        if (new RegExp('^[0-9a-fA-F]{24}$').test(params.search)) {
           condition[key] = opts.mongoose.Types.ObjectId(params.search);
           pushCondition(condition, key);
-        } catch (error) {
-          console.error(error);
+        }
+      } else if (value.instance === 'Number') {
+        if (!Number.isNaN(searchNumber)) {
```
Comment (larcin) [inline src/services/search-builder.js]: [Variables] `searchNumber` seems to be used only here. Thus you should declare your variable only in this scope (it avoids var assignment and unnecessary cast)

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,8 @@
 # Change Log
 
 ## [Unreleased]
+### Fixed
+- Search Builder - Fix broken queries on ObjectIDs and Numbers.
```
Comment (arnaudbesnier) [inline CHANGELOG.md]: Let's consider this as a feature 👍 => `Added` section
`Search builder` is very technical, we should focus on the end user experience. => "Search"
 I would add 2 lines to identify easily the 2 things you added

```
### Added
- Search - Add the search on ObjectID fields (exact match only). 
- Search - Add the search on Number fields (exact match only).
```

File: src/services/search-builder.js [hunk 5]
```diff
@@ -32,11 +29,15 @@ function SearchBuilder(model, opts, params, searchFields) {
...
+        if (new RegExp('^[0-9a-fA-F]{24}$').test(params.search)) {
           condition[key] = opts.mongoose.Types.ObjectId(params.search);
           pushCondition(condition, key);
-        } catch (error) {
-          console.error(error);
+        }
+      } else if (value.instance === 'Number') {
+        const searchNumber = Number(params.search);
+        if (!Number.isNaN(searchNumber)) {
+          condition[key] = Number(params.search);
```
Comment (arnaudbesnier) [inline src/services/search-builder.js]:
```suggestion
          condition[key] = searchNumber;
```

**Comment by adriguy** [comment]
> On search ID the `contains`, `starts_with` and `ends_with` filters does not seems to work (Actually only `is` and `is not` work)
> 
> The search also does not work on the ID.

As discussed together @larcin, the scope of making the search working on partial ID (along with making the "contains"/"starts with"/"ends with" filters working on ID) is out of the scope of this PR.
@arnaudbesnier let's create another Trello card about it?

**Comment by arnaudbesnier** [comment]
> > On search ID the `contains`, `starts_with` and `ends_with` filters does not seems to work (Actually only `is` and `is not` work)
> > The search also does not work on the ID.
> 
> As discussed together @larcin, making the search working on partial ID (along with making the "contains"/"starts with"/"ends with" filters working on ID) is out of this PR scope.
> @arnaudbesnier let's create another Trello card about it?

I consider this as a new feature request, so the right place to put a request in ProductBoard (creating an insight). If the team receives a lot of insights about this, we'll try to do the move.

File: src/services/search-builder.js (hunk 5 above)
Comment (adriguy) [reply src/services/search-builder.js]: 👎 Woooh sorry about that!

File: CHANGELOG.md (hunk 4 above)
Comment (adriguy) [reply CHANGELOG.md]: Ok, got your point! Thanks 💯

**Pull request merged**